qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -0,0 +1,160 @@
+
+            q_0: ──■──
+                 ┌─┴─┐
+            q_1: ┤ X ├
+                 └───┘
+            q_2: ─────
+
+        is represented by a 3x3 linear matrix
+
+        .. math::
+
+                \begin{pmatrix}
+                    1 & 0 & 0 \\
+                    1 & 1 & 0 \\
+                    0 & 0 & 1
+                \end{pmatrix}
+
+        """
+        if not isinstance(linear, (list, np.ndarray, QuantumCircuit)):
+            raise CircuitError(
+                "A linear function must be represented either by a list, "
+                "a numpy array, or a quantum circuit with linear gates."
+            )
+
+        if isinstance(linear, QuantumCircuit):
+            # The following function will raise a CircuitError if there are nonlinear gates.
+            self._linear = _linear_quantum_circuit_to_mat(linear)
+
+        else:
+            # Normalize to numpy array (coercing entries to 0s and 1s)
+            ok = True
+            try:
+                self._linear = np.array(linear, dtype=bool, copy=True)
+            except ValueError:
+                ok = False
+
+            if not ok:
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Check that the matrix is square
+            if not len(self._linear.shape) == 2 or self._linear.shape[0] != self._linear.shape[1]:
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Optionally, check that the matrix is invertible
+            if validate_input:
+                det = np.linalg.det(linear) % 2
+                if not np.allclose(det, 1):
+                    raise CircuitError(
+                        "A linear function must be represented by an invertible matrix."
+                    )
+
+        super().__init__(name="linear_function", num_qubits=len(self._linear), params=[])
+
+    def _define(self):
+        """Populates self.definition with a decomposition of this gate."""
+        self.definition = self.synthesize()
+
+    def synthesize(self):
+        """Synthesize ``qiskit.circuit.library.LinearFunction``.
+
+        Returns:
+            QuantumCircuit: A circuit implementing the evolution.
+        """
+        from qiskit.transpiler.synthesis import cnot_synth
+
+        return cnot_synth(self._linear)
+
+    @property
+    def linear(self):
+        """Returns the n x n matrix representing this linear function"""
+        return self._linear
+
+    def is_permutation(self) -> bool:
+        """Returns whether this linear function is a permutation."""
+        ones = np.ones([self._linear.shape[0]], dtype=int)
+        perm = np.all(np.sum(self._linear, axis=0) == ones) and np.all(
+            np.sum(self._linear, axis=1) == ones
+        )
+        return perm
+
+    def get_permutation_pattern(self):
+        """In the case that this linear function is a permutation, returns the permutation pattern"""
```

Comment:
Would be helpful to document that this method first checks if the linear function is a permutation and raises a `CircuitError` if not (so the user doesn't need to first call `is_permutation`.